Add MC reconstruction for GH+GRMHD

nilsdeppe opened the pull request:
## Proposed changes

Depends on:
- [x] #3629 

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Add MC reconstruction to GH+GRMHD

Does MC for GRMHD and degree 2 unlimited reconstruction for the metric

nilsdeppe commented:
Rebased on develop, no longer dependent

yoonso0-0 commented on src/Evolution/Systems/GrMhd/GhValenciaDivClean/FiniteDifference/MonotisedCentral.hpp:
```diff
@@ -0,0 +1,144 @@
+
+  using options = tmpl::list<>;
+  static constexpr Options::String help{
+      "Monotised central reconstruction scheme using primitive variables."};
```
Please mention metric variables in this help string

yoonso0-0 commented on src/Evolution/Systems/GrMhd/GhValenciaDivClean/FiniteDifference/MonotisedCentral.hpp:
```diff
@@ -0,0 +1,144 @@
+      gsl::not_null<std::array<Variables<TagsList>, dim>*> vars_on_lower_face,
+      gsl::not_null<std::array<Variables<TagsList>, dim>*> vars_on_upper_face,
+      const Variables<hydro::grmhd_tags<DataVector>>& volume_prims,
+      const Variables<typename System::variables_tag::type::tags_list>&
```
This argument type is not included in `reconstruction_argument_tags`

yoonso0-0 commented on src/Evolution/Systems/GrMhd/GhValenciaDivClean/FiniteDifference/MonotisedCentral.hpp:
```diff
@@ -0,0 +1,144 @@
+  void reconstruct_fd_neighbor(
+      gsl::not_null<Variables<TagsList>*> vars_on_face,
+      const Variables<hydro::grmhd_tags<DataVector>>& subcell_volume_prims,
+      const Variables<tmpl::list<
```
(same as above) This argument type is not included in reconstruction_argument_tags

yoonso0-0 commented on src/Evolution/Systems/GrMhd/GhValenciaDivClean/FiniteDifference/MonotisedCentral.hpp:
```diff
@@ -0,0 +1,144 @@
+ */
+class MonotisedCentralPrim : public Reconstructor {
+ private:
+  using tags_for_reconstruction =
```
unused type alias

yoonso0-0 commented on src/Evolution/Systems/GrMhd/GhValenciaDivClean/FiniteDifference/MonotisedCentral.hpp:
```diff
@@ -0,0 +1,144 @@
+  size_t ghost_zone_size() const override { return 2; }
+
+  using reconstruction_argument_tags =
+      tmpl::list<::Tags::Variables<hydro::grmhd_tags<DataVector>>,
```
include `"DataStructures/VariablesTag.hpp"`

yoonso0-0 commented on src/Evolution/Systems/GrMhd/GhValenciaDivClean/FiniteDifference/MonotisedCentral.hpp:
```diff
@@ -0,0 +1,144 @@
+#include "DataStructures/Tensor/TypeAliases.hpp"
+#include "Domain/Structure/MaxNumberOfNeighbors.hpp"
+#include "Domain/Tags.hpp"
+#include "Evolution/DgSubcell/Tags/Inactive.hpp"
```
not used?

yoonso0-0 commented on src/Evolution/Systems/GrMhd/GhValenciaDivClean/FiniteDifference/MonotisedCentral.hpp:
```diff
@@ -0,0 +1,144 @@
+      const FixedHashMap<
+          maximum_number_of_neighbors(dim) + 1,
+          std::pair<Direction<dim>, ElementId<dim>>,
+          evolution::dg::subcell::NeighborData,
```
Need forward decl for `evolution::dg::subcell::NeighborData`

yoonso0-0 commented on src/Evolution/Systems/GrMhd/GhValenciaDivClean/FiniteDifference/MonotisedCentral.hpp:
```diff
@@ -0,0 +1,144 @@
+      const Variables<hydro::grmhd_tags<DataVector>>& subcell_volume_prims,
+      const Variables<tmpl::list<
+          gr::Tags::SpacetimeMetric<3>, GeneralizedHarmonic::Tags::Phi<3>,
+          GeneralizedHarmonic::Tags::Pi<3>>>& subcell_volume_spacetime_metric,
```
Include for `GeneralizedHarmonic::Tags`

yoonso0-0 commented on src/Evolution/Systems/GrMhd/GhValenciaDivClean/FiniteDifference/CMakeLists.txt:
```diff
@@ -14,6 +15,10 @@ spectre_target_headers(
   FiniteDifference.hpp
+  MonotisedCentral.hpp
   Reconstructor.hpp
+  ReconstructorWork.hpp
```
Set the file name to be `ReconstructWork.hpp` and `ReconstructWork.tpp` for consistency?

yoonso0-0 commented on src/Evolution/Systems/GrMhd/GhValenciaDivClean/FiniteDifference/ReconstructorWork.hpp:
```diff
@@ -0,0 +1,92 @@
+
+namespace grmhd::GhValenciaDivClean::fd {
+/*!
+ * \brief Reconstructs \f$\rho, p, u_i, B^i\f$, and \f$\Phi\f$, then computes
```
If I understood right, what is done inside this function is 

1) compute `Wv^i` from `LorentzFactor` and `SpatialVelocity`
2) Reconstruct the quantity `Wv^i` itself
3) Pass this to `grmhd::ValenciaDivClean::fd::compute_conservatives_for_reconstruction()` to compute `LorentzFactor` and `SpatialVelocity` back

so I couldn't find lines related with `u_i`('lower spatial velocity'). Did I miss something?

nilsdeppe replied:
No, the documentation is outdated. I used to reconstruct `u_i` since that's what SpEC does

yoonso0-0 commented on src/Evolution/Systems/GrMhd/GhValenciaDivClean/FiniteDifference/ReconstructorWork.tpp:
```diff
@@ -0,0 +1,354 @@
+      for (size_t i = 0; i < 3; ++i) {
+        volume_tensor.get(i) *= get(lorentz_factor);
+      }
+      volume_tensor_ptr = &volume_tensor;
```
Q) Is the data block that volume_tensor_ptr is pointing not affected after a variable volume_tensor disappears after getting out of this scope?

nilsdeppe replied:
No, because that's stored in `lorentz_factor_times_v_I` so `lorentz_factor_times_v_I` needs to be defined outside of this block

yoonso0-0 commented on src/Evolution/Systems/GrMhd/GhValenciaDivClean/FiniteDifference/ReconstructorWork.tpp:
```diff
@@ -0,0 +1,354 @@
+                                                &subcell_mesh](auto tag_v) {
+    using tag = tmpl::type_from<decltype(tag_v)>;
+    const typename tag::type* volume_tensor_ptr = nullptr;
+    Variables<tmpl::list<
```
This allocation can be moved inside the if loop below

nilsdeppe replied:
it cannot. That will lead to dangling pointer

yoonso0-0 commented on src/Evolution/Systems/GrMhd/GhValenciaDivClean/FiniteDifference/ReconstructorWork.hpp:
```diff
@@ -0,0 +1,92 @@
+    const Mesh<3>& subcell_mesh, size_t ghost_zone_size);
+
+/*!
+ * \brief Reconstructs the mass density, velocity, and pressure, then computes
```
Please match this doc with the reconstructed prims tags (mass density, pressure, lorentz factor times spatial v, magnetic field, divergence cleaning field) or with the doc above for `reconstruct_prims_work()`.

yoonso0-0 commented on src/Evolution/Systems/GrMhd/GhValenciaDivClean/FiniteDifference/ReconstructorWork.hpp:
```diff
@@ -0,0 +1,92 @@
+ */
+template <typename TagsList, typename PrimsTags, size_t ThermodynamicDim,
+          typename F0, typename F1>
+void reconstruct_fd_neighbor_work(
```
Declaration does not match with definition in the `tpp` file

yoonso0-0 commented on src/Evolution/Systems/GrMhd/GhValenciaDivClean/FiniteDifference/ReconstructorWork.tpp:
```diff
@@ -0,0 +1,354 @@
+        neighbor_data,
+    const Mesh<3>& subcell_mesh, size_t ghost_zone_size) {
+  using prim_tags_for_reconstruction =
+      tmpl::list<hydro::Tags::RestMassDensity<DataVector>,
```
include `Utilities/TMPL.hpp`

yoonso0-0 commented on src/Evolution/Systems/GrMhd/GhValenciaDivClean/FiniteDifference/ReconstructorWork.tpp:
```diff
@@ -0,0 +1,354 @@
+    Variables<tmpl::list<
+        hydro::Tags::LorentzFactorTimesSpatialVelocity<DataVector, 3>>>
+        lorentz_factor_times_v_I{};
+    if constexpr (std::is_same_v<tag,
```
include `<type_traits>`

yoonso0-0 commented on src/Evolution/Systems/GrMhd/GhValenciaDivClean/FiniteDifference/ReconstructorWork.tpp:
```diff
@@ -0,0 +1,354 @@
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/FixedHashMap.hpp"
+#include "DataStructures/Index.hpp"
+#include "DataStructures/Tensor/EagerMath/DotProduct.hpp"
```
Not used?

yoonso0-0 commented on src/Evolution/Systems/GrMhd/GhValenciaDivClean/FiniteDifference/ReconstructorWork.tpp:
```diff
@@ -0,0 +1,354 @@
+      spacetime_vars_for_grmhd(make_not_null(&gsl::at(*vars_on_upper_face, i)));
+    }
+
+    ValenciaDivClean::fd::compute_conservatives_for_reconstruction(
```
include `Evolution/Systems/GrMhd/ValenciaDivClean/FiniteDifference/ReconstructWork.tpp` (Not sure why this function is only defined in the tpp but not in the corresponding `hpp` file)

yoonso0-0 commented on src/Evolution/Systems/GrMhd/GhValenciaDivClean/FiniteDifference/ReconstructorWork.tpp:
```diff
@@ -0,0 +1,354 @@
+                 hydro::Tags::LorentzFactorTimesSpatialVelocity<DataVector, 3>,
+                 hydro::Tags::MagneticField<DataVector, 3>,
+                 hydro::Tags::DivergenceCleaningField<DataVector>>;
+  using spacetime_tags = tmpl::list<gr::Tags::SpacetimeMetric<3>,
```
include for `gr` and `GeneralizedHarmonic` tags

yoonso0-0 commented on src/Evolution/Systems/GrMhd/GhValenciaDivClean/FiniteDifference/MonotisedCentral.cpp:
```diff
@@ -0,0 +1,289 @@
+#undef INSTANTIATION
+#undef TAGS_LIST
+#undef THERMO_DIM
+#undef DIM
```
This directive can be deleted

yoonso0-0 commented on src/Evolution/Systems/GrMhd/GhValenciaDivClean/FiniteDifference/MonotisedCentral.cpp:
```diff
@@ -0,0 +1,289 @@
+         const Index<dim>& ghost_data_extents,
+         const Direction<dim>& local_direction_to_reconstruct) {
+        ::fd::reconstruction::reconstruct_neighbor<
+            Side::Lower,
```
include `Domain/Structure/Side.hpp`

yoonso0-0 commented on src/Evolution/Systems/GrMhd/GhValenciaDivClean/FiniteDifference/MonotisedCentral.cpp:
```diff
@@ -0,0 +1,289 @@
+#define THERMO_DIM(data) BOOST_PP_TUPLE_ELEM(0, data)
+#define TAGS_LIST(data)                                                       \
+  tmpl::list<                                                                 \
+      gr::Tags::SpacetimeMetric<3>, GeneralizedHarmonic::Tags::Pi<3>,         \
```
include for `GeneralizedHarmonic`. `hydro` and `gr` tags

yoonso0-0 commented on src/Evolution/Systems/GrMhd/GhValenciaDivClean/FiniteDifference/MonotisedCentral.cpp:
```diff
@@ -0,0 +1,289 @@
+        auto& sqrt_det_spatial_metric =
+            get<gr::Tags::SqrtDetSpatialMetric<>>(*vars_on_face_ptr);
+
+        determinant_and_inverse(make_not_null(&sqrt_det_spatial_metric),
```
include `DataStructures/Tensor/EagerMath/DeterminantAndInverse.hpp`

yoonso0-0 commented on src/Evolution/Systems/GrMhd/GhValenciaDivClean/FiniteDifference/MonotisedCentral.cpp:
```diff
@@ -0,0 +1,289 @@
+      hydro::Tags::Pressure<DataVector>,                                      \
+      hydro::Tags::SpecificEnthalpy<DataVector>,                              \
+      hydro::Tags::LorentzFactorTimesSpatialVelocity<DataVector, 3>,          \
+      ::Tags::Flux<ValenciaDivClean::Tags::TildeD, tmpl::size_t<3>,           \
```
include `DataStructures/DataBox/Prefixes.hpp`

yoonso0-0 commented on tests/Unit/Evolution/Systems/GrMhd/GhValenciaDivClean/FiniteDifference/Test_MonotisedCentral.cpp:
```diff
@@ -0,0 +1,27 @@
+#include "Helpers/Evolution/Systems/GrMhd/GhValenciaDivClean/FiniteDifference/PrimReconstructor.hpp"
+
+SPECTRE_TEST_CASE(
+    "Unit.Evolution.Systems.GrMhd.GhValenciaDivClean.Fd.MonotisedCentralPrim",
```
Add test for creation, serialization, and move semantics?

yoonso0-0 commented on tests/Unit/Helpers/Evolution/Systems/GrMhd/GhValenciaDivClean/FiniteDifference/PrimReconstructor.hpp:
```diff
@@ -0,0 +1,458 @@
+#include "Utilities/Gsl.hpp"
+#include "Utilities/TMPL.hpp"
+
+#include "Parallel/Printf.hpp"
```
Unused print

yoonso0-0 commented on tests/Unit/Helpers/Evolution/Systems/GrMhd/GhValenciaDivClean/FiniteDifference/PrimReconstructor.hpp:
```diff
@@ -0,0 +1,458 @@
+void test_prim_reconstructor_impl(
+    const size_t points_per_dimension,
+    const Reconstructor& derived_reconstructor,
+    // const ComputeGrmhdSpacetimeVarsFromReconstructedSpacetimeTags&
```
Any reason for leaving this line for future use?

yoonso0-0 commented on tests/Unit/Helpers/Evolution/Systems/GrMhd/GhValenciaDivClean/FiniteDifference/PrimReconstructor.hpp:
```diff
@@ -0,0 +1,458 @@
+    const auto spatial_metric =
+        gr::spatial_metric(get<gr::Tags::SpacetimeMetric<3>>(volume_cons_vars));
+
+    // Metric is identity at cell center
```
I didn't understand this statement. Wasn't spacetime metric set to be non-Euclidean at the line 175?

nilsdeppe commented:
@yoonso0-0 I just pushed a fixup. Thanks for the review! :)

yoonso0-0 commented on src/Evolution/Systems/GrMhd/GhValenciaDivClean/FiniteDifference/MonotisedCentral.cpp:
```diff
@@ -30,7 +33,9 @@
 #include "PointwiseFunctions/GeneralRelativity/SpatialMetric.hpp"
+#include "PointwiseFunctions/GeneralRelativity/Tags.hpp"
 #include "PointwiseFunctions/Hydro/EquationsOfState/EquationOfState.hpp"
+#include "PointwiseFunctions/Hydro/Tags.hpp"
```
`#include "Evolution/Systems/GeneralizedHarmonic/Tags.hpp"`

yoonso0-0 reviewed (commented):
Fixup LGTM :) you're welcome!  Please squash with the one `include` quick fix above

nilsdeppe commented:
Rebased and squashed. Thanks for the review :)

kidder commented on src/Evolution/Systems/GrMhd/GhValenciaDivClean/FiniteDifference/MonotisedCentral.cpp:
```diff
@@ -0,0 +1,294 @@
+            local_direction_to_reconstruct);
+      },
+      [](const auto vars_on_face_ptr) {
+        // td<decltype(vars_on_face_ptr)> aeu;
```
remove

kidder commented on src/Evolution/Systems/GrMhd/GhValenciaDivClean/FiniteDifference/MonotisedCentral.hpp:
```diff
@@ -0,0 +1,144 @@
+ * ::fd::reconstruction::monotised_central() for details) and unlimited 3rd
+ * order (degree 2 polynomial) reconstruction on the metric variables.
+ *
+ * Only the spacetime metric is reconstructed when us and the neighboring
```
when we

kidder commented on src/Evolution/Systems/GrMhd/GhValenciaDivClean/FiniteDifference/MonotisedCentral.hpp:
```diff
@@ -0,0 +1,144 @@
+ * order (degree 2 polynomial) reconstruction on the metric variables.
+ *
+ * Only the spacetime metric is reconstructed when us and the neighboring
+ * element in the direction is doing FD. If a we are doing DG and a neighboring
```
are doing FD.  If we

kidder commented on src/Evolution/Systems/GrMhd/GhValenciaDivClean/FiniteDifference/MonotisedCentral.hpp:
```diff
@@ -0,0 +1,144 @@
+ * element in the direction is doing FD. If a we are doing DG and a neighboring
+ * element is doing FD, then the spacetime metric, \f$\Phi_{iab}\f$, and
+ * \f$\Pi_{ab}\f$ are all reconstructed since the Riemann solver on the DG
+ * element also needs to solver for the metric variables.
```
to solve

kidder commented on src/Evolution/Systems/GrMhd/GhValenciaDivClean/FiniteDifference/ReconstructWork.tpp:
```diff
@@ -0,0 +1,356 @@
+
+    vars_in_neighbor_count += number_of_variables;
+  });
+  tmpl::for_each<SpacetimeTagsToReconstruct>(
```
Does the order in which I reconstruct the hydro and spacetime variables matter?

nilsdeppe replied:
No because the reconstruction doesn't use things like spatial metric, etc. It just needs to be done before computing the conserved variables.

kidder reviewed (changes requested):
You can squash. I think this also will need a rebase because of #3693

nilsdeppe commented:
Thanks for the reviews! I've rebased, adjusted code to resolve conflicts that weren't git conflicts (NeighborData being removed), and squashed in the changes.

kidder commented on src/Evolution/Systems/GrMhd/GhValenciaDivClean/FiniteDifference/MonotisedCentral.hpp:
```diff
@@ -0,0 +1,141 @@
+ * order (degree 2 polynomial) reconstruction on the metric variables.
+ *
+ * Only the spacetime metric is reconstructed when we and the neighboring
+ * element in the direction are doing FD. If a we are doing DG and a neighboring
```
If we

kidder reviewed (changes requested):
small typo

nilsdeppe commented:
Rebased and squashed in wording change
